File: backend/app/schemas/lease.py
import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.models.lease import LeaseType, PaymentMethod

RentCallRuleT = Literal["contractuelle", "calendrier"]
PaymentFrequencyT = Literal[
    "mensuelle", "bimestrielle", "trimestrielle", "semestrielle", "annuelle"
]


# ── Sous-schémas pour les relations chargées ───────────────────────────────────


class TenantInLease(BaseModel):
    id: uuid.UUID
    full_name: str
    email: str | None = None
    phone: str | None = None
    national_id: str | None = None

    model_config = {"from_attributes": True}


class PropertyInLease(BaseModel):
    id: uuid.UUID
    name: str
    full_address: str

    model_config = {"from_attributes": True}


# ── CRUD schemas ───────────────────────────────────────────────────────────────


class LeaseCreate(BaseModel):
    property_id: uuid.UUID
    tenant_id: uuid.UUID
    # Co-titulaires secondaires (le principal est tenant_id)
    secondary_tenant_ids: list[uuid.UUID] = Field(default_factory=list)
    lease_type: LeaseType = LeaseType.VIDE
    start_date: date
    end_date: date | None = None
    notice_date: date | None = None
    rent_amount: float = Field(..., gt=0)
    charges_amount: float = Field(0.0, ge=0)
    deposit_amount: float = Field(0.0, ge=0)
    payment_day: int = Field(1, ge=1, le=28)
    payment_method: PaymentMethod = PaymentMethod.VIREMENT
    rent_call_rule: RentCallRuleT = "calendrier"
    payment_frequency: PaymentFrequencyT = "mensuelle"
    apl_amount: float | None = Field(None, ge=0)
    apl_tiers_payant: bool = False
    has_guarantor: bool = False
    guarantor_name: str | None = None
    guarantor_email: str | None = None
    guarantor_phone: str | None = None
    notes: str | None = None


class LeaseUpdate(BaseModel):
    # Le locataire principal peut être réassigné en modification (pas le bien).
    tenant_id: uuid.UUID | None = None
    lease_type: LeaseType | None = None
    # Si fourni, remplace la liste des co-titulaires secondaires
    secondary_tenant_ids: list[uuid.UUID] | None = None
    # Date d'entrée modifiable (correction de saisie / bail à venir).
    start_date: date | None = None
    end_date: date | None = None
    notice_date: date | None = None
    rent_amount: float | None = Field(None, gt=0)
    charges_amount: float | None = Field(None, ge=0)
    # Date d'effet d'une modification du loyer/charges (défaut : 1er du mois suivant).
    # Le mois en cours n'est pas impacté ; l'ancien montant est conservé en historique.
    rent_effective_date: date | None = None
    deposit_amount: float | None = Field(None, ge=0)
    payment_day: int | None = Field(None, ge=1, le=28)
    payment_method: PaymentMethod | None = None
    rent_call_rule: RentCallRuleT | None = None
    payment_frequency: PaymentFrequencyT | None = None
    apl_amount: float | None = Field(None, ge=0)
    apl_tiers_payant: bool | None = None
    has_guarantor: bool | None = None
    guarantor_name: str | None = None
    guarantor_email: str | None = None
    guarantor_phone: str | None = None
    notes: str | None = None


class LeaseTerminate(BaseModel):
    end_date: date
    notice_date: date | None = None


class LeaseResponse(BaseModel):
    id: uuid.UUID
    property_id: uuid.UUID
    tenant_id: uuid.UUID
    lease_type: LeaseType
    start_date: date
    end_date: date | None = None
    notice_date: date | None = None
    rent_amount: float
    charges_amount: float
    deposit_amount: float
    payment_day: int
    payment_method: PaymentMethod
    rent_call_rule: str = "calendrier"
    payment_frequency: str = "mensuelle"
    apl_amount: float | None = None
    apl_tiers_payant: bool
    has_guarantor: bool
    guarantor_name: str | None = None
    guarantor_email: str | None = None
    guarantor_phone: str | None = None
    is_active: bool
    notes: str | None = None
    total_monthly: float
    net_rent: float
    # Relations
    tenant: TenantInLease | None = None
    co_tenants: list[TenantInLease] = Field(default_factory=list)
    all_tenant_names: str | None = None
    parent_property: PropertyInLease | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class LeaseListItem(BaseModel):
    id: uuid.UUID
    property_id: uuid.UUID
    tenant_id: uuid.UUID
    tenant_full_name: str
    property_name: str
    owner_name: str | None = None
    lease_type: str
    start_date: date
    end_date: date | None = None
    rent_amount: float
    charges_amount: float
    is_active: bool
    apl_tiers_payant: bool


class LeaseListResponse(BaseModel):
    items: list[LeaseListItem]
    total: int
    skip: int
    limit: int

File: frontend/src/types/lease.ts
export type LeaseType = 'vide' | 'meuble' | 'mobilite' | 'commercial'
export type PaymentMethod = 'virement' | 'cheque' | 'prelevement' | 'especes'
export type RentCallRule = 'contractuelle' | 'calendrier'
export type PaymentFrequency =
  | 'mensuelle'
  | 'bimestrielle'
  | 'trimestrielle'
  | 'semestrielle'
  | 'annuelle'

export const RENT_CALL_RULE_LABELS: Record<RentCallRule, string> = {
  contractuelle: 'Période contractuelle',
  calendrier: 'Période calendrier',
}

export const PAYMENT_FREQUENCY_LABELS: Record<PaymentFrequency, string> = {
  mensuelle: 'Mensuelle',
  bimestrielle: 'Bimestrielle',
  trimestrielle: 'Trimestrielle',
  semestrielle: 'Semestrielle',
  annuelle: 'Annuelle',
}

export const LEASE_TYPE_LABELS: Record<LeaseType, string> = {
  vide: 'Location vide',
  meuble: 'Location meublée',
  mobilite: 'Bail mobilité',
  commercial: 'Bail commercial',
}

export const PAYMENT_METHOD_LABELS: Record<PaymentMethod, string> = {
  virement: 'Virement bancaire',
  cheque: 'Chèque',
  prelevement: 'Prélèvement',
  especes: 'Espèces',
}

export interface TenantInLease {
  id: string
  full_name: string
  email?: string
  phone?: string
  national_id?: string
}

export interface PropertyInLease {
  id: string
  name: string
  full_address: string
}

export interface Lease {
  id: string
  property_id: string
  tenant_id: string
  lease_type: LeaseType
  start_date: string
  end_date?: string
  notice_date?: string
  rent_amount: number
  charges_amount: number
  deposit_amount: number
  payment_day: number
  payment_method: PaymentMethod
  rent_call_rule: RentCallRule
  payment_frequency: PaymentFrequency
  apl_amount?: number
  apl_tiers_payant: boolean
  has_guarantor: boolean
  guarantor_name?: string
  guarantor_email?: string
  guarantor_phone?: string
  is_active: boolean
  total_monthly: number
  net_rent: number
  notes?: string
  tenant?: TenantInLease
  co_tenants?: TenantInLease[]
  all_tenant_names?: string
  parent_property?: PropertyInLease
  created_at: string
  updated_at: string
}

export interface LeaseListItem {
  id: string
  property_id: string
  tenant_id: string
  tenant_full_name: string
  property_name: string
  owner_name?: string | null
  lease_type: LeaseType
  start_date: string
  end_date?: string
  rent_amount: number
  charges_amount: number
  is_active: boolean
  apl_tiers_payant: boolean
}

export interface LeaseListResponse {
  items: LeaseListItem[]
  total: number
  skip: number
  limit: number
}
